src/main/scala/com/databricks/labs/overwatch/pipeline/SilverTransforms.scala

```diff
@@ -1314,4 +1314,119 @@ trait SilverTransforms extends SparkSessionWrapper {
 
   }
 
+  private def warehouseBase(auditRawDF: DataFrame): DataFrame = {
+    val warehouse_id_gen_w = Window.partitionBy('organization_id, 'warehouse_name)
+      .orderBy('timestamp).rowsBetween(Window.currentRow, 1000)
+    val warehouse_name_gen_w = Window.partitionBy('organization_id, 'warehouse_id)
+      .orderBy('timestamp).rowsBetween(Window.currentRow, 1000)
+    val warehouse_id_gen = first('warehouse_id, true).over(warehouse_id_gen_w)
+    val warehouse_name_gen = first('warehouse_name, true).over(warehouse_name_gen_w)
+
+    val warehouseSummaryCols = auditBaseCols ++ Array[Column](
+      when('actionName === "createEndpoint" || 'actionName === "createWarehouse",
+        get_json_object($"response.result", "$.id"))
+        .when(('actionName =!= "createEndpoint" && 'id.isNull)
+          || ('actionName =!= "createWarehouse" && 'id.isNull), 'warehouseId)
+        .otherwise('id).alias("warehouse_id"),
+      'name.alias("warehouse_name"),
+      'cluster_size,
+      'min_num_clusters.cast("long"),
+      'max_num_clusters.cast("long"),
+      'auto_stop_mins.cast("long"),
+      'spot_instance_policy,
+      'enable_photon.cast("boolean"),
+      'channel,
+      'tags,
+      'enable_serverless_compute.cast("boolean"),
+      'warehouse_type
+    )
+
+    val warehouseRaw = auditRawDF
+      .filter('serviceName === "databrickssql")
+      .selectExpr("*", "requestParams.*").drop("requestParams", "Overwatch_RunID")
+      .select(warehouseSummaryCols: _*)
+      .withColumn("warehouse_id", warehouse_id_gen)
+      .withColumn("warehouse_name", warehouse_name_gen)
+
+    val warehouseWithStructs = warehouseRaw
+      .withColumn("channel", SchemaTools.structFromJson(spark, warehouseRaw, "channel"))
+      .withColumn("tags", SchemaTools.structFromJson(spark, warehouseRaw, "tags"))
+      .scrubSchema
+
+    warehouseWithStructs
+        .withColumn("tags", SchemaTools.structToMap(warehouseWithStructs, "tags"))
+  }
+
+  protected def buildWarehouseSpec(
+                                  bronze_warehouse_snap: PipelineTable,
+                                  isFirstRun: Boolean,
+                                  untilTime: TimeTypes
+                                )(df: DataFrame): DataFrame = {
```

Comment:
do me a favor and refactor this like the workflowsTransforms. I want a dbsqlTransforms.scala object that contains all the transforms. This should have been the pattern to begin with.

This will allow you to do something like:

```scala
deriveWarehouseBase(df)
  .transform(initializeFirstRuns(df1, df2...))
  .transform(...)...
```

Reply:
done